File: flutter/sample_ui/lib/src/model/use_smileid_sample_environment.dart
/// Sandbox or production; there is no third case, and the SDK resolves to the same two hosts.
enum UseSmileIDSampleEnvironment {
  /// The sandbox, which every automated run reads.
  sandbox('sandbox', 'Sandbox', 'testapi.smileidentity.com'),

  /// Production.
  production('production', 'Production', 'api.smileidentity.com');

  const UseSmileIDSampleEnvironment(this.id, this.label, this.host);

  /// The id the result card publishes.
  final String id;

  /// The human-readable name.
  final String label;

  /// The API host the token's `api_url` claim names.
  final String host;

  /// Trailing slash, which is the form the SDK's own constants take.
  String get baseUrl => 'https://$host/';
}

/// A token's `api_url` onto an environment, matched on the parsed host: a real claim carries a
/// `/v3` path and no trailing slash, so a whole-string compare misses, and misses silently.
UseSmileIDSampleEnvironment? environmentFor(String? apiUrl) {
  final String? host = apiUrlHost(apiUrl);
  if (host == null) {
    return null;
  }
  return UseSmileIDSampleEnvironment.values
      .where((UseSmileIDSampleEnvironment it) => it.host == host)
      .firstOrNull;
}

/// The host an `api_url` names, so a rejection can say which one it saw.
String? apiUrlHost(String? apiUrl) {
  final String trimmed = apiUrl?.trim() ?? '';
  if (trimmed.isEmpty) {
    return null;
  }
  final String? host = Uri.tryParse(trimmed)?.host;
  return host == null || host.isEmpty ? null : host.toLowerCase();
}

File: expo/sample-ui/src/model/use-smile-id-sample-environment.ts
import { smileIDSampleEnvironments, type UseSmileIDSampleEnvironment } from './use-smile-id-sample-result';

/// The host each environment names.
export const smileIDSampleEnvironmentHosts: Readonly<Record<UseSmileIDSampleEnvironment, string>> = {
  sandbox: 'testapi.smileidentity.com',
  production: 'api.smileidentity.com',
};

/// What the simulated-scan chips read.
export const smileIDSampleEnvironmentLabels: Readonly<Record<UseSmileIDSampleEnvironment, string>> = {
  sandbox: 'Sandbox',
  production: 'Production',
};

/// Trailing slash, the form the SDK's own constants take.
export const smileIDSampleEnvironmentBaseUrl = (environment: UseSmileIDSampleEnvironment): string =>
  `https://${smileIDSampleEnvironmentHosts[environment]}/`;

// Scheme then authority, as java.net.URI reads it.
const AUTHORITY = /^[a-z][a-z0-9+.-]*:\/\/([^/?#]*)/i;
const HOST = /^[a-z0-9.-]+$/i;

/// The host an `api_url` names, or null.
export const smileIDSampleApiUrlHost = (apiUrl: string | null | undefined): string | null => {
  const trimmed = apiUrl?.trim() ?? '';
  // Whitespace inside is where java.net.URI throws.
  if (trimmed.length === 0 || /\s/.test(trimmed)) return null;
  const authority = AUTHORITY.exec(trimmed)?.[1];
  if (authority === undefined) return null;
  const host = authority.slice(authority.lastIndexOf('@') + 1).replace(/:\d*$/, '');
  return HOST.test(host) ? host.toLowerCase() : null;
};

/// A token's `api_url` onto an environment, by its parsed host.
export const smileIDSampleEnvironmentFor = (
  apiUrl: string | null | undefined,
): UseSmileIDSampleEnvironment | null => {
  const host = smileIDSampleApiUrlHost(apiUrl);
  return smileIDSampleEnvironments.find((environment) => smileIDSampleEnvironmentHosts[environment] === host) ?? null;
};
